VDB Merge

danrbailey opened the pull request:
This PR extends `tools/merge.h` to add a new `SumMergeOp` for parallel breadth-first summing of multiple aligned VDBs using the DynamicNodeManager. The implementation carefully preserves the order of operations to ensure the same result in spite of the non-associativity of floating-point addition. As an example, collapsing a constant value leaf node into a tile of one of the input grids will not change the resulting merged grid.

This also adds a new VDB Merge SOP which partially implements the prototype from #623. At present, the only two options available on this SOP are for merging SDF VDBs and fog volume VDBs. This implementation works well in parallel when merging a few very large VDBs and a lot of very small VDBs or some combination of both and can deep-copy and/or resample concurrently with the merging of other grids.

jmlait commented on openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc:
```diff
@@ -0,0 +1,682 @@
+    parms.add(hutil::ParmFactory(PRM_STRING, "collation", "Collation")
+        .setDefault("nameclassandtype")
+        .setChoiceListItems(PRM_CHOICELIST_SINGLE, {
+            "nameclassandtype", "Name, VDB Class and Value Type",
+            "classandtype",     "VDB Class and Value Type"
```
Since we are doing germancase here I'm not sure why we have the embedded and
I'd prefer
nameclasstype
classtype
Or maybe since these are tokens anyways,
name_class_type
class_type

jmlait replied:
One trivial collation type is "Primitive Number".   Ie, collate the first primitive of all inputs, then the second, etc.  Useful perhaps when people are lazy about naming.

danrbailey replied:
Yes, good idea. I took a quick look at doing this and thought it was going to be much more complicated to implement without some significant refactoring then realized I could simply store the primitive number in a hash map keyed by the GEO_PrimVDB::getUniqueId().

After some hesitation around this parameter, I decided it might be better to drop the "VDB Class and Value Type" part for each option and simply make that something that applies to each collation, so it now reads:
* Name
* Primitive Number
* None

If and when I implement something that can merge grids of different value types (ie merging a float grid with a double grid, etc), then this parameter will probably need to change, but this is not a trivial thing to implement anyway so best to keep it simple at this stage I think.

jmlait commented on openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc:
```diff
@@ -0,0 +1,682 @@
+            "Grids with different names will be merged if name is not part"
+            " of the collation criteria"));
+
+    // Menu of resampling options
+    parms.add(hutil::ParmFactory(PRM_STRING, "resample", "Resample VDBs")
+        .setDefault("first")
+        .setChoiceListItems(PRM_CHOICELIST_SINGLE, {
+            "first",        "To Match First VDB"
+        })
+        .setTypeExtended(PRM_TYPE_JOIN_PAIR)
+        .setDocumentation("Specify which grid to use as reference when resampling."));
+
```
Do we need to add this now if it doesn't have any options yet?  Isn't it safe just to not have it; and add it with First as the default when we have more than one choice?

danrbailey replied:
Fair point, I've made this an obsolete parameter for the time being.

jmlait commented on openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc:
```diff
@@ -0,0 +1,682 @@
+        .setDocumentation(
+            "Merge operation for Fog VDBs.\n\n"
+            "None:\n"
+            "   Add grids to the output of the SOP without merging them.\n\n"
```
Weird to have two meanings of Add here.
"Passes the grids to the output without affecting them"
perhaps?
"Leaves input fog grids unchanged"

jmlait commented on openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc:
```diff
@@ -0,0 +1,682 @@
+            "None:\n"
+            "   Add grids to the output of the SOP without merging them.\n\n"
+            "Add:\n"
+            "    Generate the sum of all fog VDBs with same collation.\n\n"));
```
Perhaps... "...fog VDBS within the same collation."

jmlait commented on openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc:
```diff
@@ -0,0 +1,682 @@
+\n\
+Unlike the VDB Combine SOP, it provides a multi-input so can merge more than two grids without needing an \
+additional merge SOP. \n\
+Float and double VDB SDF grids can be merged, all other VDB grids are added to the output without merging. \n\
```
Again, i don't see it as "adding to the output" so much as the node leaving them unchanged.  (Technically, adding to  the output is what we are doing, but people have many mental models of what actually happens :>)

danrbailey commented:
@jmlait - thanks for the feedback, lots of good improvements. I've addressed all your comments now, would be good to know what you think about these edits.

danrbailey commented on openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc:
```diff
@@ -0,0 +1,712 @@
+        .setChoiceListItems(PRM_CHOICELIST_SINGLE, {
+            "name",             "Name",
+            "primitive_number", "Primitive Number",
+            "none",             "None"
```
Yes, I think you're right, "All" is more intuitive for this use case. I can change that.
